packages/cozy-realtime/src/RetryManager.js

```diff
@@ -0,0 +1,142 @@
+  }
+
+  /**
+   * On a success
+   *
+   * If `timeBeforeSuccessfull` if not null, it will wait this amount of
+   * time without errors before acknowleding a success
+   */
+  onSuccess() {
+    if (timeBeforeSuccessfull > 0) {
+      this.startSuccessWatcher(
+        this.onSuccessAcknowledged,
+        timeBeforeSuccessfull
+      )
+    } else {
+      this.onSuccessAcknowledged()
+    }
+  }
+
+  /**
+   * Acknowledge a success
+   *
+   * @private
+   */
+  onSuccessAcknowledged() {
+    this.emit('success')
+    this.reset()
+  }
+
+  /**
+   * Clears the success watcher
+   *
+   * @private
+   */
+  clearSuccessWatcher() {
+    if (this.successWatcher) {
+      global.clearTimeout(this.successWatcher)
+      this.successWatcher = null
+    }
+  }
+
+  /**
+   * Starts the success watcher - acknowledges the success if not interrupted after some time
+   *
+   * @private
+   * @param {function} ack - will be called after the wait time
+   * @param {number} time - waiting time
+   */
+  startSuccessWatcher(ack, time) {
+    if (!this.successWatcher) {
+      const callback = () => {
+        this.successWatcher = null
+        ack()
+      }
+      this.successWatcher = global.setTimeout(callback, time)
+    }
+  }
+
+  /**
+   * On failure
+   *
+   * If we where waiting for a success timeout, it will cancel the wait.
+   *
+   * An application error (failed authentication, failed subscription) will
+   * not trigger a failure (trying to reconnect with the same parameters
+   * has no meaning anyways).
+   *
+   * @param {object} error
+   */
+  onFailure(error) {
+    this.clearSuccessWatcher()
+    this.increaseFailureCounter()
+    this.emit('failure', error)
+    if (this.hasReachedErrorLevel()) this.emit('error', error)
+  }
+
+  /**
+   * Returns if we reached a level where we need to raise an error
+   *
+   * @returns {boolean}
+   */
+  hasReachedErrorLevel() {
+    return this.retries > 0 && this.retries % raiseErrorAfterAttempts == 0
+  }
+
+  /**
+   * Reset failures counters (do not wait for before trying to connect next time)
+   */
+  reset() {
+    this.clearSuccessWatcher()
+    this.retries = 0
+    this.wait = false
+  }
+
+  /**
+   * Increase the failure counter (wait more before next attempt)
+   *
+   * @private
+   */
+  increaseFailureCounter() {
+    if (this.wait === false) {
+      // first reconnect is immediate
+      this.wait = 0
+    } else if (this.wait === 0) {
+      // base value
+      this.wait = baseWaitAfterFirstFailure
+    } else {
+      this.wait = Math.min(this.wait * 2, maxWaitBetweenRetries)
+    }
+    this.retries = this.retries + 1
+  }
+
+  /**
+   * Wait an amount of time before the next attempt (if needed)
+   */
+  async waitBeforeNextAttempt() {
+    if (this.wait) {
+      await new Promise(resolve => {
+        global.setTimeout(resolve, this.wait)
+      })
+    }
+  }
+}
+
+MicroEE.mixin(RetryManager)
+
+export default RetryManager
```

Comment:
If I'm not mistaken we already have a similar mechanism in cozy-note. Should we extract this kind of logic and make them accessible from cozy-libs or something?

Reply:
Yep, that's why I made anexternal class instead of an embeddedlogic as in notes. I still have  to check what's needed to use the same class for both. I prefer to make it fully external in a second step